<?xml version="1.0"?>
<launch>
  <!-- Basics -->
  <arg name="mav_name" default="crazyflie"/>
  <arg name="world_name" default="basic"/>
  <arg name="debug" default="false"/>
  <arg name="gui" default="true"/>
  <arg name="paused" default="true"/>
  <arg name="verbose" default="true"/>

  <!-- Crazyflie opts -->
  <arg name="enable_logging" default="true"/>
  <arg name="enable_parameters" default="true"/>
  <arg name="use_ros_time" default="true"/>
  <arg name="enable_logging_imu" default="false"/>
  <arg name="enable_logging_temperature" default="false"/>
  <arg name="enable_logging_magnetic_field" default="false"/>
  <arg name="enable_logging_pressure" default="false"/>
  <arg name="enable_logging_battery" default="false"/>
  <arg name="enable_logging_packets" default="false"/>

  <!-- MULTI-DRONE -->
  <arg name="nbQuads" default="2"/>
  <arg name="cfPrefix" default="cf"/>
  <arg name="uri" default="INADDR_ANY://19950"/>
  <arg name="handlerName" default="handler1"/>

  <!-- cf1 -->
  <arg name="tf_prefix_1" default="$(arg cfPrefix)1"/>
  <arg name="color_prop_front_1" default="Blue"/>
  <arg name="color_prop_back_1" default="Blue"/>
  <arg name="x_1" default="1.0"/>
  <arg name="y_1" default="0.0"/>
  <arg name="z_1" default="0.03"/>

  <!-- cf2 -->
  <arg name="tf_prefix_2" default="$(arg cfPrefix)2"/>
  <arg name="color_prop_front_2" default="Red"/>
  <arg name="color_prop_back_2" default="Red"/>
  <arg name="x_2" default="-1.0"/>
  <arg name="y_2" default="0.0"/>
  <arg name="z_2" default="0.03"/>

  <!-- Generic logs -->
  <rosparam>
    genericLogTopics: ["local_position","target_position"]
    genericLogTopicFrequencies: [10,10]
    genericLogTopic_local_position_Variables: ["stateEstimate.x","stateEstimate.y","stateEstimate.z","stabilizer.roll","stabilizer.pitch","stabilizer.yaw"]
    genericLogTopic_target_position_Variables: ["posCtl.targetX","posCtl.targetY","posCtl.targetZ","controller.roll","controller.pitch","controller.yaw"]
  </rosparam>

  <!-- Gazebo -->
  <env name="GAZEBO_MODEL_PATH" value="${GAZEBO_MODEL_PATH}:$(find crazyflie_gazebo)/models"/>
  <env name="GAZEBO_RESOURCE_PATH" value="${GAZEBO_RESOURCE_PATH}:$(find crazyflie_gazebo)/models"/>
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find crazyflie_gazebo)/worlds/$(arg world_name).world"/>
    <arg name="debug" value="$(arg debug)"/>
    <arg name="paused" value="$(arg paused)"/>
    <arg name="gui" value="$(arg gui)"/>
    <arg name="verbose" value="$(arg verbose)"/>
  </include>

  <!-- Spawn cf1 -->
  <group ns="$(arg cfPrefix)1">
    <include file="$(find crazyflie_gazebo)/launch/spawn_mav.launch">
      <arg name="mav_name" value="$(arg mav_name)"/>
      <arg name="model" value="$(find crazyflie_gazebo)/models/rotors_description/urdf/$(arg mav_name)_base.xacro"/>
      <arg name="tf_prefix" value="$(arg tf_prefix_1)"/>
      <arg name="color_prop_front" default="$(arg color_prop_front_1)"/>
      <arg name="color_prop_back" default="$(arg color_prop_back_1)"/>
      <arg name="x" default="$(arg x_1)"/>
      <arg name="y" default="$(arg y_1)"/>
      <arg name="z" default="$(arg z_1)"/>
    </include>
  </group>

  <!-- Spawn cf2 -->
  <group ns="$(arg cfPrefix)2">
    <include file="$(find crazyflie_gazebo)/launch/spawn_mav.launch">
      <arg name="mav_name" value="$(arg mav_name)"/>
      <arg name="model" value="$(find crazyflie_gazebo)/models/rotors_description/urdf/$(arg mav_name)_base.xacro"/>
      <arg name="tf_prefix" value="$(arg tf_prefix_2)"/>
      <arg name="color_prop_front" default="$(arg color_prop_front_2)"/>
      <arg name="color_prop_back" default="$(arg color_prop_back_2)"/>
      <arg name="x" default="$(arg x_2)"/>
      <arg name="y" default="$(arg y_2)"/>
      <arg name="z" default="$(arg z_2)"/>
    </include>
  </group>

  
 <!-- Handler para los dos drones -->
  <group ns="handler1">
    <include file="$(find crazyflie_gazebo)/launch/spawn_plugin.launch">
      <arg name="plugin_name" value="handler1" />
      <arg name="nbQuads" value="2" />              <!-- ManejarÃ¡ los dos -->
      <arg name="cfPrefix" value="cf" />            <!-- cf1 y cf2 -->
      <arg name="firstIndex" value="1" />           <!-- Empezar desde cf1 -->
      <arg name="uri" value="INADDR_ANY://19950" /> <!-- Un solo puerto -->
    
    <!-- logging/params -->
      <arg name="enable_logging" value="$(arg enable_logging)"/>
      <arg name="enable_parameters" value="$(arg enable_parameters)"/>
      <arg name="use_ros_time" value="$(arg use_ros_time)"/>
      <arg name="enable_logging_imu" value="$(arg enable_logging_imu)"/>
      <arg name="enable_logging_temperature" value="$(arg enable_logging_temperature)"/>
      <arg name="enable_logging_magnetic_field" value="$(arg enable_logging_magnetic_field)"/>
      <arg name="enable_logging_pressure" value="$(arg enable_logging_pressure)"/>
      <arg name="enable_logging_battery" value="$(arg enable_logging_battery)"/>
      <arg name="enable_logging_packets" value="$(arg enable_logging_packets)"/>
    </include>
  </group>


</launch>

